source/util/bit_stream.h

```diff
@@ -0,0 +1,371 @@
+  const size_t block_num = val >> block_exponent;
+  const size_t pos = GetLowerBits(val, block_exponent);
+  if (block_num & 1) {
+    // Negative.
+    return -1LL - ((block_num >> 1) << block_exponent) - pos;
+  } else {
+    // Positive.
+    return ((block_num >> 1) << block_exponent) + pos;
+  }
+}
+
+// Converts |buffer| to a stream of zeroes and ones.
+template <typename T>
+std::string BufferToStream(const std::vector<T>& buffer) {
+  std::stringstream ss;
+  for (auto it = buffer.begin(); it != buffer.end(); ++it) {
+    std::string str = std::bitset<sizeof(T) * 8>(*it).to_string();
+    // Strings generated by std::bitset::to_string are read right to left.
+    // Reversing to left to right.
+    std::reverse(str.begin(), str.end());
+    ss << str;
+  }
+  return ss.str();
+}
+
+// Converts a left-to-right input string to a buffer of |T| words.
+template <typename T>
+std::vector<T> StreamToBuffer(std::string str) {
+  // The input string is left-to-right, the input argument of std::bitset needs
+  // to right-to-left. Instead of reversing tokens, reverse the entire string
+  // and iterate tokens from end to begin.
+  std::reverse(str.begin(), str.end());
+  const size_t word_size = sizeof(T) * 8;
+  std::vector<T> buffer;
+  buffer.reserve(NumBitsToNumWords<sizeof(T)>(str.length()));
+  for (ssize_t index = str.length() - word_size; index >= 0;
+       index -= word_size) {
+    buffer.push_back(static_cast<T>(std::bitset<sizeof(T) * 8>(
+        str, index, word_size).to_ulong()));
+  }
+  const size_t suffix_length = str.length() % word_size;
+  if (suffix_length != 0) {
+    buffer.push_back(static_cast<T>(std::bitset<sizeof(T) * 8>(
+        str, 0, suffix_length).to_ulong()));
+  }
+  return buffer;
+}
+
+// Adds '0' chars at the end of the string until the size is a multiple of N.
+template <size_t N>
+inline std::string PadToWord(std::string&& str) {
+  const size_t tail_length = str.size() % N;
+  if (tail_length != 0)
+    str += std::string(N - tail_length, '0');
+  return str;
+}
+
+// Adds '0' chars at the end of the string until the size is a multiple of N.
+template <size_t N>
+inline std::string PadToWord(const std::string& str) {
+  return PadToWord<N>(std::string(str));
+}
+
+// Converts a left-to-right stream of bits to std::bitset.
+template <size_t N>
+inline std::bitset<N> StreamToBitset(std::string str) {
+  std::reverse(str.begin(), str.end());
+  return std::bitset<N>(str);
+}
+
+// Converts first |num_bits| of std::bitset to a left-to-right stream of bits.
+template <size_t N>
+inline std::string BitsetToStream(const std::bitset<N>& bits, size_t num_bits = N) {
+  std::string str = bits.to_string().substr(N - num_bits);
+  std::reverse(str.begin(), str.end());
+  return str;
+}
+
+// Converts a left-to-right stream of bits to uint64.
+inline uint64_t StreamToBits(std::string str) {
+  std::reverse(str.begin(), str.end());
+  return std::bitset<64>(str).to_ullong();
```

Comment:
I'm not sure if this to_ullong method is availble on Android.  Will need to test.